### ISSUE TITLE
Remove user fails in case of linked federated identity

### ISSUE THREAD

**Issue by bjoernhype** [keycloak/keycloak#29531]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

storage

### Describe the bug

Deletion of a user with federated identity / user federation fails.

Keycloak log (debug) shows:
```
2024-05-08 09:39:26,235 DEBUG [org.hibernate.SQL] (executor-thread-5) delete from USER_ENTITY where ID=?

2024-05-08 09:39:26,237 DEBUG [WebApplicationException] (executor-thread-5) Restarting handler chain for exception exception: org.keycloak.services.ErrorResponseException: HTTP 500 Internal Server Error

        at org.keycloak.services.ErrorResponse.error(ErrorResponse.java:43)

        at org.keycloak.services.ErrorResponse.error(ErrorResponse.java:36)

        at org.keycloak.services.resources.admin.UserResource.deleteUser(UserResource.java:632)

        at org.keycloak.services.resources.admin.UserResource$quarkusrestinvoker$deleteUser_1ab1e247046f32b48fb55de7f71f5540f6c8ea14.invoke(Unknown Source)

        at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)

        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)

        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)

        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)

        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)

        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)

        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)

        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)

        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)

        at java.base/java.lang.Thread.run(Unknown Source)

2024-05-08 09:39:26,238 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (executor-thread-5) JtaTransactionWrapper rollback
```

Deletion of this user suddenly works if I set user_entity#federation_link of this user to null (by DB tool).

I looked at the current implementation, and saw in JpaUserProvider:
```
    private void removeUser(UserEntity user) {
        String id = user.getId();
        em.createNamedQuery("deleteUserRoleMappingsByUser").setParameter("user", user).executeUpdate();
        em.createNamedQuery("deleteUserGroupMembershipsByUser").setParameter("user", user).executeUpdate();
        em.createNamedQuery("deleteUserConsentClientScopesByUser").setParameter("user", user).executeUpdate();
        em.createNamedQuery("deleteUserConsentsByUser").setParameter("user", user).executeUpdate();

        em.remove(user);
        em.flush();
    }
```
First several dependencies of the user (foreign keys) are deleted, and than the user itself is deleted.
My guess is that there is missing one line to remove the user from `federated_identity`, as I get a
```
ERROR: update or delete on table "user_entity" violates foreign key constraint "fk404288b92ef007a6" on table "federated_identity"
  Detail: Key (id)=(fc3f954c-f359-402d-b978-21c8e167bfef) is still referenced from table "federated_identity".
```
when doing the steps from code manually. 

### Version

24.0.4

### Regression

- [X] The issue is a regression. It worked with Keycloak 20 and several older versions. 21-23 we did not test.

### Expected behavior

User can be deleted

### Actual behavior

Deletion of user with federated identity fails: 
KC admin UI shows a "The user could not be deleted User couldn't be deleted" on try to delete the user.
KC REST API returns a 400.

### How to Reproduce?

* Create a user with user federation / federated identity. In my scenario, I have both SAML and user federation active, the user exists in both, and I do first log in by SAML. Keycloak than creates the user with links to SAML **and** user federation. This seems ok for me. 
* Try to delete this user by either admin UI or REST API

**Comment by bjoernhype** [comment]
I tried with custom user storage and without SAML identity provider now. Than there is no problem!

So it seems that when using SAML and also detecting & linking custom user storage, something is done differently in user creation, which prevents the user from deletion. Perhaps on deletion there is a check whether an entry in federated_identity should also be deleted, and caused by the concrete creation, this is not deleted, which results in the user deletion to fails because of the FK at federated_identity

**Comment by sguilhen** [comment]
@bjoernhype I'm having trouble reproducing this - when I setup a broker and a storage provider that both have the same user, and I login via broker, I get a message saying user exists (it finds user with same username in the storage provider) and asks me to link the user.. then I have to login using the storage provider, and the user is created as a `federated_user` linked to the broker via `broker_link` table.

Inspecting this user I see both the federation link to the storage provider, and the idp link. Removing the user causes no issues at all and I see nothing added to the `federated_identity` table.

I can only get the user added to this table when there are no enabled storage providers in the realm - then when I log in via broker the user is imported as a local user (visible in the `user_entity` table) and we have an entry in the `federated_identity` table. Removing this user again works fine.

So, when having a storage provider enabled, I was never able to get the user imported as a `federated_identity` - it always is created as a `federated_user`, so when it is deleted there's nothing else to delete in the `federated_identity` table (and even if somehow there is something there to delete, the relationship has a CASCADE_REMOVE, that should take care of it when deleting the user entity).

I guess my question is how exactly does your DB look like after you perform a login of a a user that exists both in the IdP and storage provider for the first time, assuming you do so via IdP? What do you see in the `federated_identity`, `federated_user`, `user_entity` and `broker_link` tables look like for this new user? Does the login process inform you of an existing user and asks you to link the information?

**Comment by sguilhen** [comment]
Also, a few more considerations:

You mentioned that setting the `federation_link` in the user to NULL in the DB fixed it - this essentially signals to Keycloak that the user is a local one, so it won't invoke the custom storage provider when deleting, which kinda points in the direction that the custom provider is failing to delete the user. Are you able to debug your custom provider to see what's happening on delete?

Also, following the steps to delete users manually in the DB will indeed show FK issues because it doesn't account for the cascade delete in the user entity that handles that step, so the issue has probably nothing to do with that.

**Comment by bjoernhype** [comment]
Thanks for inspecting this. I see its not easy to reproduce this, so I had a look in the Keycloak code:

In UserStorageManager#removeUser there is
```
        StorageId storageId = new StorageId(user.getId());

        if (storageId.getProviderId() == null) {
            String federationLink = user.getFederationLink();
            boolean linkRemoved = federationLink == null || Optional.ofNullable(
                    getStorageProviderInstance(realm, federationLink, UserRegistrationProvider.class))
                    .map(provider -> provider.removeUser(realm, user))
                    .orElse(false);

            return localStorage().removeUser(realm, user) && linkRemoved;
        }
```
As `user.getId()` doesn't start with `f:`, we enter the `if`.
In my special scenario, 1. the user has a federation_link, and 2. I did not implement the optional `UserRegistrationProvider` in my `UserStorageProvider`, because its optional and docs say Keycloak will handle the user deletion.
As consequence, `linkRemoved` will be `false`, and `removeUser` returns `false`, which will trigger an exception in `UserResource#deleteUser` .
If setting `federation_link` in the user to `NULL` manually, `linkRemoved` will be `true` instead, because of the first part of the condition.
So this seems to be a problem. Must I implement `UserRegistrationProvider` to be able to remove such a user from Keycloak, or is this a bug in Keycloak impl?
My use case is that I first delete the user in my software, and after that call the Keycloak delete user REST-API endpoint, so I thought there is nothing to impl for user deletion in my `UserStorageProvider`.
I don't know if this is my only problem, but it seems a part of it.
Debugging Keycloak or my `UserStorageProvider` isn't trivial for me, as I currently start them in a docker scenario.

Another point:
I see a named query `deleteFederatedIdentityByUser` in `FederatedIdentityEntity`, but its used nowhere.
The FK from `federated_identity.user_id` to `user_entity` doesn't have a `on delete` .
In `JpaUserProvider#removeUser` I don't see a cascade, too.
=> From view of Keycloak code, there seems a problem.

**Comment by pedroigor** [comment]
@bjoernhype It makes more sense to implement `UserRegistrationProvider` so that you indicate the user was previously removed. By doing that, you avoid the exception being thrown when calling `UserResource.delete` and the user should be deleted as expected.

I'm also wondering why you delete the user first in your software and then you call the User API. You probably have a constraint/reason that is forcing you to do this but I would like to understand more.

In summary, if implementing the `UserRegistrationProvider` interface works for you, I'm in favor of closing the issue.

**Comment by bjoernhype** [comment]
Hi @pedroigor , thanks for your response.
I just tried, and yes: Implementing `UserRegistrationProvider` works for me :)
So I still think it would be better to return `true` in the `orElse` (s.o.), but for me it is ok to close this, now.

The reason for deleting in our software instead of UserRegistrationProvider is mainly historic. I'll think about if we change this sometime.

Thanks for your effort
